query/src/storages/fuse/operations/mutation/compact_mutator.rs

```diff
@@ -0,0 +1,164 @@
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+use std::sync::Arc;
+
+use common_exception::Result;
+use opendal::Operator;
+
+use super::block_filter::all_the_columns_ids;
+use crate::sessions::QueryContext;
+use crate::storages::fuse::io::BlockCompactor;
+use crate::storages::fuse::io::BlockWriter;
+use crate::storages::fuse::io::MetaReaders;
+use crate::storages::fuse::io::SegmentWriter;
+use crate::storages::fuse::io::TableMetaLocationGenerator;
+use crate::storages::fuse::meta::Location;
+use crate::storages::fuse::meta::SegmentInfo;
+use crate::storages::fuse::meta::Statistics;
+use crate::storages::fuse::meta::TableSnapshot;
+use crate::storages::fuse::statistics::reducers::reduce_block_metas;
+use crate::storages::fuse::statistics::reducers::reduce_statistics;
+use crate::storages::fuse::FuseTable;
+
+pub struct CompactMutator<'a> {
+    ctx: &'a Arc<QueryContext>,
+    location_generator: &'a TableMetaLocationGenerator,
+    base_snapshot: &'a TableSnapshot,
+    data_accessor: Operator,
+    row_per_block: usize,
+    block_per_seg: usize,
+    segments: Vec<Location>,
+    summarys: Vec<Statistics>,
+}
+
+impl<'a> CompactMutator<'a> {
+    pub fn try_create(
+        ctx: &'a Arc<QueryContext>,
+        location_generator: &'a TableMetaLocationGenerator,
+        base_snapshot: &'a TableSnapshot,
+        row_per_block: usize,
+        block_per_seg: usize,
+    ) -> Result<Self> {
+        let data_accessor = ctx.get_storage_operator()?;
+        Ok(Self {
+            ctx,
+            location_generator,
+            base_snapshot,
+            data_accessor,
+            row_per_block,
+            block_per_seg,
+            segments: Vec::new(),
+            summarys: Vec::new(),
+        })
+    }
+
+    pub async fn into_new_snapshot(self) -> Result<(TableSnapshot, String)> {
+        let snapshot = self.base_snapshot;
+        let mut new_snapshot = TableSnapshot::from_previous(snapshot);
+        new_snapshot.segments = self.segments.clone();
+        // update the summary of new snapshot
+        let new_summary = reduce_statistics(&self.summarys)?;
+        new_snapshot.summary = new_summary;
+
+        // write the new segment out (and keep it in undo log)
+        let snapshot_loc = self.location_generator.snapshot_location_from_uuid(
+            &new_snapshot.snapshot_id,
+            new_snapshot.format_version(),
+        )?;
+        let bytes = serde_json::to_vec(&new_snapshot)?;
+        self.data_accessor
+            .object(&snapshot_loc)
+            .write(bytes)
+            .await?;
```

Comment:
`object(..).write(..)` may failed for some transient io errors,  maybe we should use `io::write_meta` instead?

https://github.com/datafuselabs/databend/blob/main/query/src/storages/fuse/operations/mutation/mutations_collector.rs#L138-L142